src/librustc_mir/transform/generator.rs

```diff
@@ -1072,14 +1072,27 @@ fn create_generator_resume_function<'tcx>(
     use rustc::mir::interpret::PanicInfo::{
         GeneratorResumedAfterPanic,
         GeneratorResumedAfterReturn,
+        AsyncResumedAfterReturn,
+        AsyncResumedAfterPanic,
     };
 
     // Jump to the entry point on the unresumed
     cases.insert(0, (UNRESUMED, BasicBlock::new(0)));
-    // Panic when resumed on the returned state
-    cases.insert(1, (RETURNED, insert_panic_block(tcx, body, GeneratorResumedAfterReturn)));
-    // Panic when resumed on the poisoned state
-    cases.insert(2, (POISONED, insert_panic_block(tcx, body, GeneratorResumedAfterPanic)));
+
+    // Panic when resumed on the returned or poisoned state
+    match body.generator_kind {
+        Some(GeneratorKind::Async(_)) => {
+            cases.insert(1, (RETURNED, insert_panic_block(tcx, body, AsyncResumedAfterReturn)));
```

Comment:
With my suggestion about a single variant, this code would not be duplicated

Reply:
Ok I'll give it a shot. Makes sense. Cheers

Reply:
Done